lightning-invoice/src/payment.rs

```diff
@@ -211,13 +232,29 @@ pub trait Router<S: Score> {
 	) -> Result<Route, LightningError>;
 }
 
-/// Number of attempts to retry payment path failures for an [`Invoice`].
-///
-/// Note that this is the number of *path* failures, not full payment retries. For multi-path
-/// payments, if this is less than the total number of paths, we will never even retry all of the
-/// payment's paths.
+/// Strategy for retrying payment path failures for an [`Invoice`].
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
-pub struct RetryAttempts(pub usize);
+pub enum Retry {
+	/// Count Out. Max number of attempts to retry payment.
+	///
+	/// Note that this is the number of *path* failures, not full payment retries. For multi-path
+	/// payments, if this is less than the total number of paths, we will never even retry all of the
+	/// payment's paths.
+	Attempts(usize),
+	#[cfg(feature = "std")]
+	/// Time Out. Max time spent to retry payment
+	Timeout(Duration),
+}
+
+impl Retry {
+	fn is_retryable_now(&self, attempts: &PaymentAttempts) -> bool {
+		match self {
+			Retry::Attempts(0) => false,
+			Retry::Attempts(max_retry_count) => max_retry_count + 1 > attempts.counts,
+			Retry::Timeout(max_duration) => *max_duration >= SystemTime::now().duration_since(attempts.first_attempted_at).unwrap()
+		}
+	}
```

Comment:
This is more of a curiosity, but do we only ever care about the current time, or could this be used to ask if something is retry-able at other times?

```suggestion
	fn is_retryable_now(&self, attempts: &PaymentAttempts, at: SystemTime) -> bool {
		match self {
			Retry::Attempts(0) => false,
			Retry::Attempts(max_retry_count) => max_retry_count + 1 > attempts.counts,
			Retry::Timeout(max_duration) => *max_duration >= at.duration_since(attempts.first_attempted_at).unwrap()
		}
	}
```

Reply:
yea that was my initial reason adding the `latest_attempted_at` but since this is not needed anywhere yet so I removed it according to other feedback.